lib/engine/game/g_22_mars/game.rb

```diff
@@ -168,18 +168,103 @@ class Game < Game::Base
           },
         ].freeze
 
+        # Game specific constants
         LAST_OR = 11
+        FIRST_REVOLT_CHECK_OR = 7
+        LAST_REVOLT_CHECK_OR = 10
 
-        @or = 0
+        def setup
+          @or = 0
+          @revolt_round = -1
+          @three_or_round = false
+        end
 
         def end_now?(_after)
           @or == LAST_OR
         end
 
+        def setup_preround
+          revolts, permits = @companies.partition(&:revolt?)
+          sorted_permits = permits.sort_by { rand }
+          sorted_revolts = revolts.sort_by { rand }
+
+          @companies = sorted_permits + sorted_revolts
+
+          @companies.drop(@players.size).each_with_index do |company, index|
+            if index >= @corporations.size
+              company.close!
+              @log << "#{company.name} is discarded"
+            else
+              corporation = @corporations[index]
+              company.owner = corporation
+              corporation.companies << company
+              @log << "#{company.name} is assigned to #{corporation.name}"
+            end
+
+            @companies.delete(company)
+          end
+        end
+
+        def unowned_purchasable_companies
+          @companies
+        end
+
+        def init_round
+          G22Mars::Round::PermitPurchase.new(self, [G22Mars::Step::SelectOrDiscard])
+        end
+
         def new_operating_round(round_num = 1)
           @or += 1
 
-          super
+          if @or == 9
+            @operating_rounds = 3
+            @three_or_round = true
+          end
+
+          round = super
+
+          check_revolt!
+
+          round
+        end
+
+        def or_round_finished
+          # In case we get phase change during the last OR set we ensure we have 3 ORs
+          @operating_rounds = 3 if @three_or_round
+        end
+
+        def revolt_happened?
+          @revolt_happened
+        end
+
+        def check_revolt!
+          return if @or < FIRST_REVOLT_CHECK_OR || revolt_happened?
+
+          # Generate a random number 0 to 3 (2, 1) to simulate 1 in 4 (in OR7, 1 in 3 in OR8 etc) cube draw chance
+          revolt_check = (rand % (LAST_REVOLT_CHECK_OR - @or + 1))
```

Comment:
Is the timing of the revolt going to be determined by the game seed? If there are other things that call `rand` between operating rounds then this might be non-deterministic, if `rand` could be called a different number of times. An alternative would be to determine the order of cube draws at the start of the game.